Add markup as designed (#574)
unicode-org/message-format-wg: spec/data-model/README.md

## Patterns

Each `Pattern` represents a linear sequence, without selectors.
Each element of the `body` array MUST either be a non-empty string or an `Expression` or `Markup` object.

> String values represent literal _text_,

Author: ```suggestion
String values represent literal _text_.
```

> while `Expression` and `Markup` wrap each of the potential _expression_ and _markup_ shapes.

Author: ```suggestion
`Expression` wraps each of the potential _expression_ shapes.
`Markup` wraps each of the potential _markup_ shapes.
```

> while `Expression` and `Markup` wrap each of the potential _expression_ and _markup_ shapes.

Author: Applied as two separate suggestions, to work around GitHub limitations.

The `body` strings are the "cooked" _text_ values, i.e. escape sequences are processed.

> Implementations MUST NOT rely on the set of `Expression` and `Markup` interfaces being exhaustive,
> as future versions of this specification MAY define additional expressions.

Reviewer: I think the term "expression" is overloaded here. Also, the second normative word shouldn't be normative (it's permitting a future version of this specification to do something). Perhaps:

```suggestion
Implementations MUST NOT rely on the set of `Expression` and 
`Markup` interfaces defined in this document being exhaustive.
Future versions of this specification might define additional
types of expression or markup.
```

## Patterns

Each `Pattern` represents a linear sequence, without selectors.

> Each element of the `body` array MUST be a non-empty string, an `Expression` or a `Markup` object.

Reviewer: Nit: use the Oxford comma

```suggestion
Each element of the `body` array MUST be a non-empty string, an `Expression`, or a `Markup` object.
```

String values represent literal _text_.
`Expression` wraps each of the potential _expression_ shapes.
`Markup` wraps each of the potential _markup_ shapes.
The `body` strings are the "cooked" _text_ values, i.e. escape sequences are processed.

Implementations MUST NOT rely on the set of `Expression` and 
`Markup` interfaces defined in this document being exhaustive.
Future versions of this specification might define additional

> types of expression or markup.

Reviewer: Minor wording. "Types of" is unncessary and also misleading, since we might use the same "type" of expression for a new function (for example)

```suggestion
expressions or markup.
```

```

A `FunctionAnnotation` represents a _function_ _annotation_.

> The `name` does not include the `:` starting sigil.

Reviewer: I think there's a problem here. We're using the word `name` (here and in the interface definitions above and below). This is fine for `VariableRef`, but not for `FunctionAnnotation`, which should use `identifier` (that is, including the namespace)

Author: I've spun this off into a separate issue #575, as it's a separate concern from the markup implementation.

```ts
interface UnsupportedAnnotation {
  type: "unsupported-annotation";
  sigil: "!" | "@" | "%" | "^" | "&" | "*" | "+" | "<" | ">" | "?" | "~";
  source: string;
}
```

## Markup

> Each of _markup-open_, _markup-close_, and _markup-standalone_ are represented
> by the corresponding `MarkupOpen`, `MarkupStandalone`, and `MarkupClose`.

Reviewer: This reads oddly (both in terms of English and in terms of stylistically). The other sections start with the interface names, while this starts with what is represented. Perhaps:

```suggestion
Each of the markup interfaces `MarkupOpen`, `MarkupStandalone`,
and `MarkupClose` represent the _markup_ `type` of the corresponding
`kind`.
```

> The `name` in these does not include the starting sigils `#` and `/`.

Reviewer: `name` => `identifier` in this section

Author: See #575.

```ts
interface UnsupportedAnnotation {
  type: "unsupported-annotation";

>   sigil: "!" | "@" | "%" | "^" | "&" | "*" | "+" | "<" | ">" | "?" | "~";

Reviewer: what order is this list in? Probably we should sort the reserved sigils into code point order?

Author: I think it's currently something close to the layout order on a US keyboard? I've no opinion on the order here.

  source: string;
}
```

## Markup

A `Markup` object is either `MarkupOpen`, `MarkupStandalone`, or `MarkupClose`,

> which are differentiated by `kind`.

Reviewer: Should we call out that open/standalone/close carries no semantic meaning, unless assigned by the implementation?

Author: I don't think that necessarily belongs here in the data model, given that it's an optional part of the spec?

Reviewer: I think we should call it out, but agree that it doesn't belong here.

> The `name` in these does not include the starting sigils `#` and `/`.

Reviewer: We should call out the closer too

```suggestion
The `name` in these does not include the starting sigils `#` and `/` 
or the ending sigil `/`.
```

```ts
interface UnsupportedAnnotation {
  type: "unsupported-annotation";
  sigil: "!" | "@" | "%" | "^" | "&" | "*" | "+" | "<" | ">" | "?" | "~";
  source: string;
}
```

## Markup

> Each of the markup interfaces `MarkupOpen`, `MarkupStandalone`,
> and `MarkupClose` represent the _markup_ `type` of the corresponding
> `kind`.

Reviewer 2: I don't know how to read this sentence... what is it trying to communicate? Perhaps something like this?
```suggestion
A `Markup` object is either `MarkupOpen`, `MarkupStandalone`, or `MarkupClose`,
which are differentiated by `kind`.
```
